app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt

```diff
@@ -81,23 +109,116 @@ private fun orientation(a: Coordinates, b: Coordinates, c: Coordinates): Int {
     c.lng in minOf(a.lng, b.lng)..maxOf(a.lng, b.lng)
 
-/** Checks if a polygon formed by the given vertices is self-intersecting. */
+/** Represents a line segment defined by two endpoints. */
+private data class PolygonEdge(val a: Coordinates, val b: Coordinates)
+
+/**
+ * Returns whether the polygon represented by vertices is closed (i.e., its first and last vertices
+ * are the same).
+ */
+fun isClosed(coordinates: List<Coordinates>): Boolean =
+  coordinates.size >= 4 && coordinates.firstOrNull() == coordinates.lastOrNull()
+
+/** Checks if two segments share a common endpoint. */
+private fun shareEndpoint(s1: PolygonEdge, s2: PolygonEdge): Boolean =
+  s1.a == s2.a || s1.a == s2.b || s1.b == s2.a || s1.b == s2.b
+
+/**
+ * Builds a list of edges from the given polygon vertices.
+ *
+ * Each [PolygonEdge] connects two consecutive vertices in the list. If the polygon is closed (first
+ * vertex == last vertex), an additional edge is added from the last vertex back to the first to
+ * complete the loop.
+ *
+ * Examples:
+ * - For an open polyline [A, B, C], edges = (A→B, B→C)
+ * - For a closed polygon [A, B, C, A], edges = (A→B, B→C, C→A)
+ *
+ * @param vertices List of polygon vertices.
+ * @return A list of [PolygonEdge]s that represent the polygon’s sides.
+ */
+private fun buildSegments(vertices: List<Coordinates>): List<PolygonEdge> {
+  val n = vertices.size
+  if (n < 2) return emptyList()
+
+  val closed = isClosed(vertices)
+
+  val edges = ArrayList<PolygonEdge>(if (closed) n else n - 1)
+
+  // Create edges between each consecutive pair of vertices.
+  for (i in 0 until n - 1) {
+    edges += PolygonEdge(vertices[i], vertices[i + 1])
+  }
+
+  // If the polygon is closed, add an edge from the last vertex back to the first
+  // to ensure the polygon forms a complete loop.
+  if (closed) edges += PolygonEdge(vertices[n - 1], vertices[0])
+
+  return edges
+}
+
+/**
+ * Returns whether two edges in a polygon are adjacent.
+ *
+ * Adjacency rules:
+ * - An edge is considered adjacent to itself (useful for "skip" checks).
+ * - Consecutive indices are adjacent (e.g., 3 and 4).
+ * - If the polygon is closed, the first and last edges are also adjacent.
+ *
+ * @param firstEdgeIndex Index of the first edge in the edges list.
+ * @param secondEdgeIndex Index of the second edge in the edges list.
+ * @param edgeCount Total number of edges.
+ * @param isClosedPolygon Whether the polygon is closed (first vertex equals last vertex).
+ */
+private fun areAdjacent(
+  firstEdgeIndex: Int,
+  secondEdgeIndex: Int,
+  edgeCount: Int,
+  isClosedPolygon: Boolean,
+): Boolean {
+  if (firstEdgeIndex == secondEdgeIndex) return true
+  if (abs(firstEdgeIndex - secondEdgeIndex) == 1) return true
+  if (isClosedPolygon) {
+    val last = edgeCount - 1
+    if (
+      (firstEdgeIndex == 0 && secondEdgeIndex == last) ||
+        (secondEdgeIndex == 0 && firstEdgeIndex == last)
+    ) {
+      return true
+    }
+  }
```

Comment:
We can simplify these method (this and the other ones) by specifying exactly which geometry we are dealing with (i.e. linear ring or polygon). Handling for generic vertices is causing readability issues as you need to check for `completeness` at every step. Either pass a linear ring which ensures that the first and last vertices are same or handle generically for a list of vertices without checking for `completeness`.